src/anndata/_io/specs/methods.py

```diff
@@ -532,23 +547,38 @@ def write_vlen_string_array_zarr(
     _writer: Writer,
     dataset_kwargs: Mapping[str, Any] = MappingProxyType({}),
 ):
-    import numcodecs
-
-    if Version(numcodecs.__version__) < Version("0.13"):
-        msg = "Old numcodecs version detected. Please update for improved performance and stability."
-        warnings.warn(msg)
-        # Workaround for https://github.com/zarr-developers/numcodecs/issues/514
-        if hasattr(elem, "flags") and not elem.flags.writeable:
-            elem = elem.copy()
-
-    f.create_dataset(
-        k,
-        shape=elem.shape,
-        dtype=object,
-        object_codec=numcodecs.VLenUTF8(),
-        **dataset_kwargs,
-    )
-    f[k][:] = elem
+    if is_zarr_v2():
+        import numcodecs
+
+        if Version(numcodecs.__version__) < Version("0.13"):
+            msg = "Old numcodecs version detected. Please update for improved performance and stability."
+            warnings.warn(msg)
+            # Workaround for https://github.com/zarr-developers/numcodecs/issues/514
+            if hasattr(elem, "flags") and not elem.flags.writeable:
+                elem = elem.copy()
+
+        f.create_dataset(
+            k,
+            shape=elem.shape,
+            dtype=object,
+            object_codec=numcodecs.VLenUTF8(),
+            **dataset_kwargs,
+        )
+        f[k][:] = elem
+    else:
+        from numcodecs import VLenUTF8
+
+        dataset_kwargs = dataset_kwargs.copy()
+        if "compressor" in dataset_kwargs:
+            dataset_kwargs.pop("compressor")
+        f.create_array(
+            k,
+            shape=elem.shape,
+            dtype=str if ad.settings.zarr_write_format == 3 else object,
+            filters=[VLenUTF8()] if ad.settings.zarr_write_format == 2 else None,
```

Comment:
This is a bit confusing. why `== 3` and `== 2`? Are there more valid values?

if not, maybe

```py
dtype, filters = ... if ad.settings.zarr_write_format == 3 else ...
f.create_array(k, shape=elem.shape, dtype=dtype, filters=filters, **dataset_kwargs)
```